bridge/gitlab/import.go

```diff
@@ -0,0 +1,351 @@
+		if err != nil {
+			return fmt.Errorf("issue creation: %v", err)
+		}
+
+		// Loop over all notes
+		for gi.iterator.NextNote() {
+			note := gi.iterator.NoteValue()
+			if err := gi.ensureNote(repo, b, note); err != nil {
+				return fmt.Errorf("note creation: %v", err)
+			}
+		}
+
+		// Loop over all label events
+		for gi.iterator.NextLabelEvent() {
+			labelEvent := gi.iterator.LabelEventValue()
+			if err := gi.ensureLabelEvent(repo, b, labelEvent); err != nil {
+				return fmt.Errorf("label event creation: %v", err)
+			}
+		}
+
+		// commit bug state
+		if err := b.CommitAsNeeded(); err != nil {
+			return fmt.Errorf("bug commit: %v", err)
+		}
+	}
+
+	if err := gi.iterator.Error(); err != nil {
+		fmt.Printf("import error: %v\n", err)
+		return err
+	}
+
+	fmt.Printf("Successfully imported %d issues and %d identities from Gitlab\n", gi.importedIssues, gi.importedIdentities)
+	return nil
+}
+
+func (gi *gitlabImporter) ensureIssue(repo *cache.RepoCache, issue *gitlab.Issue) (*cache.BugCache, error) {
+	// ensure issue author
+	author, err := gi.ensurePerson(repo, issue.Author.ID)
+	if err != nil {
+		return nil, err
+	}
+
+	// resolve bug
+	b, err := repo.ResolveBugCreateMetadata(keyGitlabUrl, issue.WebURL)
+	if err != nil && err != bug.ErrBugNotExist {
+		return nil, err
+	}
+
+	// if bug was never imported
+	if err == bug.ErrBugNotExist {
+		cleanText, err := text.Cleanup(issue.Description)
+		if err != nil {
+			return nil, err
+		}
+
+		// create bug
+		b, _, err = repo.NewBugRaw(
+			author,
+			issue.CreatedAt.Unix(),
+			issue.Title,
+			cleanText,
+			nil,
+			map[string]string{
+				keyOrigin:    target,
+				keyGitlabId:  parseID(issue.ID),
+				keyGitlabUrl: issue.WebURL,
+			},
+		)
+
+		if err != nil {
+			return nil, err
+		}
+
+		// importing a new bug
+		gi.importedIssues++
+
+		return b, nil
+	}
+
+	return b, nil
+}
+
+func (gi *gitlabImporter) ensureNote(repo *cache.RepoCache, b *cache.BugCache, note *gitlab.Note) error {
+	id := parseID(note.ID)
+
+	// ensure issue author
+	author, err := gi.ensurePerson(repo, note.Author.ID)
+	if err != nil {
+		return err
+	}
+
+	hash, errResolve := b.ResolveOperationWithMetadata(keyGitlabId, id)
+	if errResolve != cache.ErrNoMatchingOp {
+		return err
+	}
+
+	noteType, body := GetNoteType(note)
+	switch noteType {
+	case NOTE_CLOSED:
+		_, err = b.CloseRaw(
+			author,
+			note.CreatedAt.Unix(),
+			map[string]string{
+				keyGitlabId:  id,
+				keyGitlabUrl: "",
+			},
+		)
+		return err
+
+	case NOTE_REOPENED:
+		_, err = b.OpenRaw(
+			author,
+			note.CreatedAt.Unix(),
+			map[string]string{
+				keyGitlabId:  id,
+				keyGitlabUrl: "",
+			},
+		)
+		return err
+
+	case NOTE_DESCRIPTION_CHANGED:
+		issue := gi.iterator.IssueValue()
+
+		// since gitlab doesn't provide the issue history
+		// we should check for "changed the description" notes and compare issue texts
+		// TODO: Check only one time and ignore next 'description change' within one issue
```

Comment:
it's still useful to have the full history, no ?

Reply:
Yes but all what Gitlab api provide is `changed the description` message and an `UpdatedAt` date. The only thing we have about each update is the `UpdatedAt`

Reply:
for each description change we compare the issue text with the local text, if they are different we update our local version and go on. any other description change in the same issue will try to do same thing and find everything up to date.